python/ray/util/client/dataclient.py

```diff
@@ -73,66 +78,157 @@ def _start_datathread(self) -> threading.Thread:
             # will be added to self.asyncio_waiting_data
 
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Gracefully shutting down
-                logger.info("Cancelling data channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from data channel")
-            else:
-                logger.exception(
-                    "Got Error from data channel -- shutting down:")
+    def _acknowledge(self, req_id: int):
+        """
+        Puts an acknowledge request on the request queue periodically.
+        Lock should be held before calling this.
+        """
+        assert self.lock.locked()
+        self._acknowledge_counter += 1
+        if self._acknowledge_counter % ACKNOWLEDGE_BATCH_SIZE == 0:
+            self.request_queue.put(
+                ray_client_pb2.DataRequest(
+                    acknowledge=ray_client_pb2.AcknowledgeRequest(
+                        req_id=req_id)))
+
+    def _reconnect_channel(self):
+        """
+        Attempts to reconnect the gRPC channel and resend outstanding
+        requests. First, the server is pinged to see if the current channel
+        still works. If the ping fails, then the current channel is closed
+        and replaced with a new one.
+
+        Once a working channel is available, a new request queue is made
+        and filled with any outstanding requests to be resent to the server.
+        """
+        # Ping the server to see if the current channel is reuseable
+        try:
+            ping_succeeded = self.client_worker.ping_server(timeout=5)
+        except grpc.RpcError:
+            ping_succeeded = False
+
+        if not ping_succeeded:
+            # Ping failed, try refreshing the data channel
+            logger.warning(
+                "Encountered connection issues in the data channel. "
+                "Attempting to reconnect.")
+            try:
+                self.client_worker._connect_channel(reconnecting=True)
+            except ConnectionError:
+                logger.warning("Failed to reconnect the data channel")
+                raise
+            logger.info("Reconnection succeeded!")
+
+        # Recreate the request queue, and resend outstanding requests
+        with self.lock:
+            self.request_queue = queue.Queue()
+            for request in self.outstanding_requests.values():
+                # Resend outstanding requests
+                self.request_queue.put(request)
 
     def close(self) -> None:
+        if self.request_queue is not None:
```

Comment:
We can avoid duplicating logic and putting `None` into the queue twice. Why not just add
```
self.request_queue.put(ray_client_pb2.DataRequest(
                connection_cleanup=ray_client_pb2.ConnectionCleanupRequest()))
```
above `self.request_queue.put(None)` below.